### PR TITLE
feat: ai-prompt-decorator plugin

### PR DESCRIPTION
**Pull request by shreemaan-abhishek**
### Description

Implement the prompt decorator plugin.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/plugins/ai-prompt-decorator.lua [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+local function get_request_body_table()
+    local body, err = core.request.get_body()
+    if not body then
+        return nil, { message = "could not get body: " .. err }
+    end
+
+    local body_tab, err = core.json.decode(body)
+    if not body_tab then
+        return nil, { message = "could not get parse JSON request body: ", err }
```
Comment (membphis) [inline apisix/plugins/ai-prompt-decorator.lua:76]:
```suggestion
        return nil, { message = "could not get parse JSON request body: " .. err }
```

File: apisix/plugins/ai-prompt-decorator.lua [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    end
+
+    return body_tab
+end
+
+
+local function decorate(conf, body_tab)
+    local new_messages = conf.prepend or {}
```
Comment (membphis) [inline apisix/plugins/ai-prompt-decorator.lua:84]: we can reuse this temporary table, eg: top level empty table

File: apisix/plugins/ai-prompt-decorator.lua [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    end
+
+    if not body_tab.messages then
+        return 400, "messages missing from request body"
+    end
+    decorate(conf, body_tab) -- will decorate body_tab in place
+
+    local new_jbody, err = core.json.encode(body_tab)
+    if not new_jbody then
+        return 500, { message = "failed to parse modified JSON request body: ", err }
```
Comment (membphis) [inline apisix/plugins/ai-prompt-decorator.lua:110]: ditto 
```suggestion
        return 500, { message = "failed to parse modified JSON request body: " .. err }
```

**Review by bzp2010 — COMMENTED** [review]
This plugin should also be added to `config.yaml.example`.

**Comment by shreemaan-abhishek** [comment]
> This plugin should also be added to `config.yaml.example`.

done

**Review by bzp2010 — APPROVED** [review]

File: docs/en/latest/plugins/ai-prompt-decorator.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      "nodes": {
+        "api.openai.com:443": 1
+      },
+      "pass_host": "node",
+      "scheme": "https"
+    }
+  }'
+```
+
+The upstream node can be any arbitrary value because it won't be contacted.
```
Comment (kayx23) [inline docs/en/latest/plugins/ai-prompt-decorator.md:84]: I understand your intention but maybe remove this sentence 🤔 reads a bit confusing

File: docs/en/latest/plugins/ai-prompt-decorator.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+## Description
+
+The `ai-prompt-decorator` plugin simplifies access to AI providers and models by appending or prepending
```
Comment (kayx23) [inline docs/en/latest/plugins/ai-prompt-decorator.md:32]:
```suggestion
The `ai-prompt-decorator` plugin simplifies access to LLM providers, such as OpenAI and Anthropic, and their models by appending or prepending
```

File: docs/en/latest/plugins/ai-prompt-decorator.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+## Plugin Attributes
+
+| **Field**         | **Type** | **Description**                                     | **Required**    |
+| ----------------- | -------- | --------------------------------------------------- | --------------- |
+| `prepend`         | Array    | An array of prompt objects to be prepended          | Conditionally\* |
+| `prepend.role`    | String   | Role of the message (`system`, `user`, `assistant`) | Yes             |
+| `prepend.content` | String   | Content of the message. Minimum length: 1           | Yes             |
+| `append`          | Array    | An array of prompt objects to be appended           | Conditionally\* |
+| `append.role`     | String   | Role of the message (`system`, `user`, `assistant`) | Yes             |
+| `append.content`  | String   | Content of the message. Minimum length: 1           | Yes             |
```
Comment (kayx23) [inline docs/en/latest/plugins/ai-prompt-decorator.md:44]: Move the column to the front?

image

**Review by kayx23 — COMMENTED** [review]
Other doc changes lgtm.

**Review by moonming — APPROVED** [review]

File: docs/en/latest/plugins/ai-prompt-decorator.md (hunk 4 above)
Comment (shreemaan-abhishek) [reply docs/en/latest/plugins/ai-prompt-decorator.md:84]: this is a key information so we should keep it, no?

Comment (shreemaan-abhishek) [reply docs/en/latest/plugins/ai-prompt-decorator.md:83]: whoops sorry this is a wrong info 😂. Good catch 🙏🏼

**Review by kayx23 — DISMISSED** [review]
doc LGTM

File: apisix/plugins/ai-prompt-template.lua [hunk 7]
```diff
@@ -72,7 +72,7 @@ local schema = {
 
 local _M = {
     version  = 0.1,
-    priority = 1060,
+    priority = 1071,
```
Comment (shreemaan-abhishek) [inline apisix/plugins/ai-prompt-template.lua:75]: `ai-prompt-decorator` plugin should run before `ai-prompt-decorator`

**Review by moonming — APPROVED** [review]

**Review by bzp2010 — APPROVED** [review]

**Review by kayx23 — APPROVED** [review]

**Pull request merged**